File: Rokt.Widget/android/src/main/java/com/rokt/reactnativesdk/RNRoktWidgetModule.java

package com.rokt.reactnativesdk;

import android.app.Activity;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.WritableMap;
import com.facebook.react.bridge.ReadableType;
import com.facebook.react.modules.core.DeviceEventManagerModule;
import com.facebook.react.uimanager.NativeViewHierarchyManager;
import com.facebook.react.uimanager.UIBlock;
import com.facebook.react.uimanager.UIManagerModule;
import com.rokt.roktsdk.Rokt;
import com.rokt.roktsdk.Rokt.RoktEventHandler;
import com.rokt.roktsdk.Widget;

import java.lang.ref.WeakReference;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.List;
import java.util.Set;
import java.util.HashSet;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

import android.content.Context;
import android.graphics.Typeface;
import android.content.res.AssetManager;
import java.lang.ref.WeakReference;

/**
 * Copyright 2020 Rokt Pte Ltd
 *
 * Licensed under the Rokt Software Development Kit (SDK) Terms of Use
 * Version 2.0 (the "License");
 *
 * You may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at https://rokt.com/sdk-license-2-0/
 */

public class RNRoktWidgetModule extends ReactContextBaseJavaModule {

    private final ReactApplicationContext reactContext;
    private RoktEventHandler roktEventHandler;
    private Boolean debug = false;
    private final Set<Typeface> typefacesSet;

    Map<Long, Rokt.RoktCallback> listeners = new LinkedHashMap<Long, Rokt.RoktCallback>() {
        @Override
        protected boolean removeEldestEntry(Map.Entry<Long, Rokt.RoktCallback> eldest) {
            int MAX_LISTENERS = 5;
            return this.size() > MAX_LISTENERS;
        }
    };

    RNRoktWidgetModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
        this.typefacesSet = new HashSet<Typeface>();
    }

    @ReactMethod
    public void initialize(String roktTagId, String appVersion, final ReadableArray fontPostScriptNames) {
        Activity currentActivity = getCurrentActivity();
        if (currentActivity != null && appVersion != null && roktTagId != null) {
            Rokt.INSTANCE.setFrameworkType(Rokt.SdkFrameworkType.ReactNative.INSTANCE);
            Rokt.INSTANCE.init(roktTagId, appVersion, currentActivity, true, readableArrayToSetOfStrings(fontPostScriptNames));
        } else {
            logDebug("Activity, roktTagId and AppVersion cannot be null");
        }
    }

    @ReactMethod
    public void execute(final String viewName, final ReadableMap attributes, final ReadableMap placeholders, final ReadableMap fontsMap) {
        if (viewName == null) {
            logDebug("Execute failed. ViewName cannot be null");
            return;
        }
        Activity currentActivity = getCurrentActivity();
        final Map<String, WeakReference<Typeface>> typefaceMap = new HashMap<>();
        if (currentActivity != null) {
            typefaceMap.putAll(retrieveFonts(readableMapToMapOfStrings(fontsMap), currentActivity));
        }

        UIManagerModule uiManager = reactContext.getNativeModule(UIManagerModule.class);
        uiManager.addUIBlock(new UIBlock() {
            @Override
            public void execute(NativeViewHierarchyManager nativeViewHierarchyManager) {
                Rokt.INSTANCE.execute(viewName, readableMapToMapOfStrings(attributes), createRoktCallback(), safeUnwrapPlaceholders(placeholders, nativeViewHierarchyManager), typefaceMap);
            }
        });
    }

    @ReactMethod
    public void execute2Step(final String viewName, final ReadableMap attributes, final ReadableMap placeholders, final ReadableMap fontsMap) {
        if (viewName == null) {
            logDebug("Execute failed. ViewName cannot be null");
            return;
        }
        Activity currentActivity = getCurrentActivity();
        final Map<String, WeakReference<Typeface>> typefaceMap = new HashMap<>();
        if (currentActivity != null) {
            typefaceMap.putAll(retrieveFonts(readableMapToMapOfStrings(fontsMap), currentActivity));
        }

        UIManagerModule uiManager = reactContext.getNativeModule(UIManagerModule.class);
        uiManager.addUIBlock(new UIBlock() {
            @Override
            public void execute(NativeViewHierarchyManager nativeViewHierarchyManager) {
                Rokt.INSTANCE.execute2Step(viewName, readableMapToMapOfStrings(attributes), createRoktCallback(), safeUnwrapPlaceholders(placeholders, nativeViewHierarchyManager), new Rokt.RoktEventCallback() {
                    @Override
                    public void onEvent(Rokt.RoktEventType roktEventType, final Rokt.RoktEventHandler roktEventHandler) {
                        setRoktEventHandler(roktEventHandler);
                        if (roktEventType == Rokt.RoktEventType.FirstPositiveEngagement) {
                            logDebug("onFirstPositiveEvent was fired");
                            sendEvent(reactContext, "FirstPositiveResponse", null);
                        }
                    }
                }, typefaceMap);
            }
        });
    }


    private void sendEvent(ReactContext reactContext,
                           String eventName,
                           @Nullable WritableMap params) {
        if (reactContext != null) {
            reactContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter.class)
                    .emit(eventName, params);
        }
    }

    @ReactMethod
    public void setFulfillmentAttributes(final ReadableMap attributes) {
        if (this.roktEventHandler != null) {
            Map<String, String> fulfillmentAttributes = readableMapToMapOfStrings(attributes);
            if (fulfillmentAttributes != null) {
                roktEventHandler.setFulfillmentAttributes(fulfillmentAttributes);
                logDebug("Calling setFulfillmentAttributes");
            } else {
                logDebug("Fulfillment attributes must be a map of Strings");
            }
        } else {
            logDebug("RoktEventHandler is null, make sure you run execute2Step before calling setFulfillmentAttributes");
        }
    }

    private void setRoktEventHandler(RoktEventHandler roktEventHandler) {
        this.roktEventHandler = roktEventHandler;
    }

    @Override
    public String getName() {
        return "RNRoktWidget";
    }

    @ReactMethod
    public void setEnvironmentToStage() {
        Rokt.INSTANCE.setEnvironment(Rokt.Environment.Stage.INSTANCE);
    }

    @ReactMethod
    public void setEnvironmentToProd() {
        Rokt.INSTANCE.setEnvironment(Rokt.Environment.Prod.INSTANCE);
    }

    @ReactMethod
    public void toggleDebug(Boolean enabled) {
        this.debug = enabled;
    }

    private void logDebug(String message) {
        if (debug) {
            Log.d("Rokt", message);
        }
    }

    private Map<String, String> readableMapToMapOfStrings(final ReadableMap attributes) {
        if (attributes != null) {
            Map<String, Object> map = attributes.toHashMap();
            Map<String, String> newMap = new HashMap<String, String>();
            for (Map.Entry<String, Object> entry : map.entrySet()) {
                if (entry.getValue() instanceof String) {
                    newMap.put(entry.getKey(), (String) entry.getValue());
                }
            }
            return newMap;
        }

        return null;
    }

    private Set<String> readableArrayToSetOfStrings(final ReadableArray array) {
        if (array == null) return null;
        return array
                .toArrayList()
                .stream()
                .filter(s -> s instanceof String)
                .map(s -> (String) s)
                .collect(Collectors.toSet());
    }

    private Rokt.RoktCallback createRoktCallback() {
        Rokt.RoktCallback callback = new Rokt.RoktCallback() {
            @Override
            public void onLoad() {
                sendCallback("onLoad", null);
            }

            @Override
            public void onUnload(@NonNull Rokt.UnloadReasons unloadReasons) {
                sendCallback("onUnLoad", unloadReasons.toString());
            }

            @Override
            public void onShouldShowLoadingIndicator() {
                sendCallback("onShouldShowLoadingIndicator", null);
            }

            @Override
            public void onShouldHideLoadingIndicator() {
                sendCallback("onShouldHideLoadingIndicator", null);
            }
        };
        listeners.put(System.currentTimeMillis(), callback);
        return callback;
    }

    private void sendCallback(final String eventValue, @Nullable final String reason) {
        WritableMap params = Arguments.createMap();
        params.putString("callbackValue", eventValue);
        if (reason != null) {
            params.putString("reason", reason);
        }
        sendEvent(reactContext, "RoktCallback", params);
    }

    private Map<String, WeakReference<Widget>> safeUnwrapPlaceholders(final ReadableMap placeholders, final NativeViewHierarchyManager nativeViewHierarchyManager) {
        final Map<String, WeakReference<Widget>> placeholderMap = new HashMap<>();

        if (placeholders != null) {
            for (Map.Entry<String, Object> entry : placeholders.toHashMap().entrySet()) {
                if (entry.getValue() instanceof Double) {
                    int tag = ((Double) entry.getValue()).intValue();
                    View view = nativeViewHierarchyManager.resolveView(tag);
                    if (view instanceof Widget) {
                        placeholderMap.put(entry.getKey(), new WeakReference(view));
                    }
                }
            }
        }

        return placeholderMap;
    }

    private Map<String, WeakReference<Typeface>> retrieveFonts(final Map<String, String> fontNameMap, final Context context) {
        final String FONTS_ASSET_PATH = "fonts/";
        try {
            AssetManager am = context.getAssets();
            Map<String, WeakReference<Typeface>> typefaceMap = new HashMap<>();
            fontNameMap.forEach((k, v) -> {
                Typeface tf = Typeface.createFromAsset(am, FONTS_ASSET_PATH + v);
                typefacesSet.add(tf);
                typefaceMap.put(k, new WeakReference<Typeface>(tf));
            });
            return typefaceMap;
        } catch (RuntimeException e) {
            return new HashMap<String, WeakReference<Typeface>>();
        }
    }
}

File: Rokt.Widget/android/src/main/java/com/rokt/reactnativesdk/RNRoktWidgetModule.kt
package com.rokt.reactnativesdk

import android.app.Activity
import android.os.Handler
import android.util.Log
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule.RCTDeviceEventEmitter
import com.facebook.react.uimanager.NativeViewHierarchyManager
import com.facebook.react.uimanager.UIManagerModule
import com.rokt.roktsdk.FulfillmentAttributes
import com.rokt.roktsdk.Rokt
import com.rokt.roktsdk.RoktConfig
import com.rokt.roktsdk.Rokt.Environment.Prod
import com.rokt.roktsdk.Rokt.RoktEventHandler
import com.rokt.roktsdk.Rokt.RoktEventType
import com.rokt.roktsdk.Rokt.SdkFrameworkType.ReactNative
import com.rokt.roktsdk.RoktEvent
import com.rokt.roktsdk.Widget
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import java.lang.ref.WeakReference

/**
 * Copyright 2024 Rokt Pte Ltd
 *
 * Licensed under the Rokt Software Development Kit (SDK) Terms of Use
 * Version 2.0 (the "License");
 *
 * You may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at https://rokt.com/sdk-license-2-0/
 */
class RNRoktWidgetModule internal constructor(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(
        reactContext
    ) {
    private var roktEventHandler: RoktEventHandler? = null
    private var fulfillmentAttributesCallback: FulfillmentAttributes? = null
    private var debug = false

    private val listeners: MutableMap<Long, Rokt.RoktCallback> =
        object : LinkedHashMap<Long, Rokt.RoktCallback>() {
            override fun removeEldestEntry(eldest: Map.Entry<Long, Rokt.RoktCallback>): Boolean =
                this.size > MAX_LISTENERS
        }

    @ReactMethod
    fun initialize(roktTagId: String?, appVersion: String?) {
        initRokt(roktTagId, appVersion, 4) { activity ->
            Rokt.init(
                roktTagId = requireNotNull(roktTagId),
                appVersion = requireNotNull(appVersion),
                activity = activity
            )
        }
    }

    @ReactMethod
    fun initializeWithFontFiles(roktTagId: String?, appVersion: String?, fontsMap: ReadableMap?) {
        initRokt(roktTagId, appVersion, 4) { activity ->
            Rokt.init(
                roktTagId = requireNotNull(roktTagId),
                appVersion = requireNotNull(appVersion),
                activity = activity,
                fontPostScriptNames = HashSet(),
                fontFilePathMap = readableMapToMapOfStrings(fontsMap)
            )
        }
    }

    @ReactMethod
    fun execute(viewName: String?, attributes: ReadableMap?, placeholders: ReadableMap?) {
        executeInternal(viewName, attributes, placeholders)
    }

    @ReactMethod
    fun executeWithConfig(
        viewName: String?,
        attributes: ReadableMap?,
        placeholders: ReadableMap?,
        roktConfig: ReadableMap?
    ) {
        executeInternal(viewName, attributes, placeholders, roktConfig)
    }

    private fun executeInternal(
        viewName: String?,
        attributes: ReadableMap?,
        placeholders: ReadableMap?,
        roktConfig: ReadableMap? = null
    ) {
        if (viewName == null) {
            logDebug("Execute failed. ViewName cannot be null")
            return
        }

        val uiManager = reactContext.getNativeModule(UIManagerModule::class.java)
        startRoktEventListener(Rokt.events(viewName), viewName)

        val config = roktConfig?.let { buildRoktConfig(it) }
        uiManager?.addUIBlock { nativeViewHierarchyManager ->
            Rokt.execute(
                viewName = viewName,
                attributes = readableMapToMapOfStrings(attributes),
                callback = createRoktCallback(),
                placeholders = safeUnwrapPlaceholders(placeholders, nativeViewHierarchyManager),
                config = config
            )
        }
    }

    @ReactMethod
    fun execute2Step(viewName: String?, attributes: ReadableMap?, placeholders: ReadableMap?) {
        execute2StepInternal(viewName, attributes, placeholders)
    }

    @ReactMethod
    fun execute2StepWithConfig(
        viewName: String?,
        attributes: ReadableMap?,
        placeholders: ReadableMap?,
        roktConfig: ReadableMap?
    ) {
        execute2StepInternal(viewName, attributes, placeholders, roktConfig)
    }

    private fun execute2StepInternal(
        viewName: String?,
        attributes: ReadableMap?,
        placeholders: ReadableMap?,
        roktConfig: ReadableMap? = null
    ) {
        if (viewName == null) {
            logDebug("Execute failed. ViewName cannot be null")
            return
        }

        val uiManager = reactContext.getNativeModule(UIManagerModule::class.java)
        startRoktEventListener(Rokt.events(viewName), viewName)
        val config = roktConfig?.let { buildRoktConfig(it) }
        uiManager?.addUIBlock { nativeViewHierarchyManager ->
            Rokt.execute2Step(
                viewName = viewName,
                attributes = readableMapToMapOfStrings(attributes),
                callback = createRoktCallback(),
                placeholders = safeUnwrapPlaceholders(placeholders, nativeViewHierarchyManager),
                roktEventCallback = object : Rokt.RoktEventCallback {
                    override fun onEvent(
                        eventType: RoktEventType,
                        roktEventHandler: RoktEventHandler
                    ) {
                        setRoktEventHandler(roktEventHandler)
                        if (eventType == RoktEventType.FirstPositiveEngagement) {
                            logDebug("onFirstPositiveEvent was fired")
                            sendEvent(reactContext, "FirstPositiveResponse", null)
                        }
                    }
                },
                config = config
            )
        }
    }

    private fun initRokt(
        roktTagId: String?,
        appVersion: String?,
        retryCount: Int,
        init: (activity: Activity) -> Unit
    ) {
        if (roktTagId == null || appVersion == null) {
            logDebug("roktTagId and appVersion cannot be null")
            return
        }
        Rokt.setFrameworkType(ReactNative)
        Rokt.globalEvents()
        startRoktEventListener(Rokt.globalEvents())
        if (currentActivity == null) {
            // When the init was called from ReactComponent init and the activity is not fully resumed,
            // the currentActivity could be null.
            // Add a delay in this scenario and call the init.
            // Recursive call initRokt to retry until retryCount becomes 0
            Handler(reactContext.mainLooper).postDelayed({
                currentActivity?.let(init) ?: run {
                    if (retryCount == 0) {
                        logDebug("Failed to initialize Rokt. Activity is null!!")
                    } else {
                        initRokt(roktTagId, appVersion, retryCount - 1, init)
                    }
                }
            }, 100)
        } else {
            currentActivity?.let(init)
        }
    }

    private fun sendEvent(
        reactContext: ReactContext?,
        eventName: String,
        params: WritableMap?
    ) {
        reactContext?.getJSModule(RCTDeviceEventEmitter::class.java)?.emit(eventName, params)
    }

    @ReactMethod
    fun setFulfillmentAttributes(attributes: ReadableMap?) {
        if (roktEventHandler != null) {
            val fulfillmentAttributes = readableMapToMapOfStrings(attributes)
            roktEventHandler?.setFulfillmentAttributes(fulfillmentAttributes)
            logDebug("Calling setFulfillmentAttributes")
        } else {
            logDebug("RoktEventHandler is null, make sure you run execute2Step before calling setFulfillmentAttributes")
        }

        if (fulfillmentAttributesCallback != null) {
            val fulfillmentAttributes = readableMapToMapOfStrings(attributes)
            fulfillmentAttributesCallback?.sendAttributes(fulfillmentAttributes)
            logDebug("Calling setFulfillmentAttributes")
        }
    }

    private fun setRoktEventHandler(roktEventHandler: RoktEventHandler) {
        this.roktEventHandler = roktEventHandler
    }

    override fun getName(): String {
        return "RNRoktWidget"
    }

    @ReactMethod
    fun setEnvironmentToStage() {
        Rokt.setEnvironment(Rokt.Environment.Stage)
    }

    @ReactMethod
    fun setEnvironmentToProd() {
        Rokt.setEnvironment(Prod)
    }

    @ReactMethod
    fun setLoggingEnabled(enabled: Boolean) {
        this.debug = enabled
        Rokt.setLoggingEnabled(enabled)
    }

    private fun logDebug(message: String) {
        if (debug) {
            Log.d("Rokt", message)
        }
    }

    private fun readableMapToMapOfStrings(attributes: ReadableMap?): Map<String, String> =
        attributes?.toHashMap()?.filter { it.value is String }?.mapValues { it.value as String }
            ?: emptyMap()


    private fun createRoktCallback(): Rokt.RoktCallback {
        val callback: Rokt.RoktCallback = object : Rokt.RoktCallback {
            override fun onLoad() {
                sendCallback("onLoad", null)
            }

            override fun onUnload(reason: Rokt.UnloadReasons) {
                sendCallback("onUnLoad", reason.toString())
            }

            override fun onShouldShowLoadingIndicator() {
                sendCallback("onShouldShowLoadingIndicator", null)
            }

            override fun onShouldHideLoadingIndicator() {
                sendCallback("onShouldHideLoadingIndicator", null)
            }
        }
        listeners[System.currentTimeMillis()] = callback
        return callback
    }

    private fun sendCallback(eventValue: String, reason: String?) {
        val params = Arguments.createMap()
        params.putString("callbackValue", eventValue)
        if (reason != null) {
            params.putString("reason", reason)
        }
        sendEvent(reactContext, "RoktCallback", params)
    }

    private fun safeUnwrapPlaceholders(
        placeholders: ReadableMap?,
        nativeViewHierarchyManager: NativeViewHierarchyManager
    ): Map<String, WeakReference<Widget>> {
        val placeholderMap: MutableMap<String, WeakReference<Widget>> = HashMap()

        if (placeholders != null) {
            placeholderMap.putAll(placeholders.toHashMap()
                .filterValues { value -> value is Double }
                .mapValues { pair -> (pair.value as Double).toInt() }
                .mapValues { pair -> nativeViewHierarchyManager.resolveView(pair.value) as? Widget }
                .filterValues { value -> value != null }
                .mapValues { WeakReference(it.value as Widget) })
        }
        return placeholderMap
    }

    private fun String.toColorMode(): RoktConfig.ColorMode {
        return when (this) {
            "dark" -> RoktConfig.ColorMode.DARK
            "light" -> RoktConfig.ColorMode.LIGHT
            else -> RoktConfig.ColorMode.SYSTEM
        }
    }

    private fun buildRoktConfig(
        roktConfig: ReadableMap?
    ): RoktConfig {
        val builder = RoktConfig.Builder()
        val configMap: Map<String, String> = readableMapToMapOfStrings(roktConfig)
        configMap["colorMode"]?.let {
            builder.colorMode(it.toColorMode())
        }

        return builder.build()
    }

    private fun startRoktEventListener(flow: Flow<RoktEvent>, viewName: String? = null) {
        (currentActivity as? LifecycleOwner)?.lifecycleScope?.launch {
            (currentActivity as LifecycleOwner).repeatOnLifecycle(Lifecycle.State.CREATED) {
                flow.collect { event ->
                    val params = Arguments.createMap()
                    var eventName = ""
                    val placementId: String? = when (event) {
                        is RoktEvent.FirstPositiveEngagement -> {
                            eventName = "FirstPositiveEngagement"
                            fulfillmentAttributesCallback = event.fulfillmentAttributes
                            event.id
                        }

                        RoktEvent.HideLoadingIndicator -> {
                            eventName = "HideLoadingIndicator"
                            null
                        }

                        is RoktEvent.OfferEngagement -> {
                            eventName = "OfferEngagement"
                            event.id
                        }

                        is RoktEvent.PlacementClosed -> {
                            eventName = "PlacementClosed"
                            event.id
                        }

                        is RoktEvent.PlacementCompleted -> {
                            eventName = "PlacementCompleted"
                            event.id
                        }

                        is RoktEvent.PlacementFailure -> {
                            eventName = "PlacementFailure"
                            event.id
                        }

                        is RoktEvent.PlacementInteractive -> {
                            eventName = "PlacementInteractive"
                            event.id
                        }

                        is RoktEvent.PlacementReady -> {
                            eventName = "PlacementReady"
                            event.id
                        }

                        is RoktEvent.PositiveEngagement -> {
                            eventName = "PositiveEngagement"
                            event.id
                        }

                        RoktEvent.ShowLoadingIndicator -> {
                            eventName = "ShowLoadingIndicator"
                            null
                        }

                        is RoktEvent.InitComplete -> {
                            eventName = "InitComplete"
                            params.putString("status", event.success.toString())
                            null
                        }
                    }

                    placementId?.let { params.putString("placementId", it) }
                    params.putString("event", eventName)
                    viewName?.let { params.putString("viewName", it) }
                    sendEvent(reactContext, "RoktEvents", params)
                }
            }
        }
    }
}

private const val MAX_LISTENERS = 5
